un-implement Request from *proto.BatchRequest

tbg opened the pull request:
and various other progress towards #2143.

tbg committed:
gc queue: avoid adding BatchRequest to client.Batch

tbg committed:
intent resolution: avoid adding BatchRequest to client.Batch

tbg committed:
change push with empty pushee

the Pushee was only used to create a priority;
now we instead create the priority at the origin
of the request and wrap it in an otherwise
empty transaction. That allows us to skip semi-
awkward random number generation in PushTxn
and also avoids a technical issue regarding
UserPriority and batch.InternalAddRequest.

tbg committed:
add PushNow to PushRequest

so we do not have to use the request timestamp,
which makes this call easier to use in circum-
stances in which the request timestamp is not
easily controlled.

This change was motivated by removing an instance
of batch.InternalAddRequest being called with a
*proto.BatchRequest with custom headers.

tbg committed:
Store.ExecuteCmd takes BatchRequest

and was renamed to `Send`, so that `Store` now is
a `client.Sender`.
Added a helper for sending a single request through
`replica.AddCmd` in anticipation of a similar goal
for `Replica`.

tbg committed:
only expose {node,db}.Batch

also, *BatchRequest is not a proto.Request any more
and Replica.AddCmd now takes a BatchRequest.
Trim some fat in proto/{api,batch}.go.

tbg committed:
fix bug in auto-txn wrap

this used to be done only for Range ops, but
a Batch can span ranges no matter what (for
example two Put operations which hit different
ranges).
An associated test needed updates because it
wrote two values in the same batch, so they
were written at the same walltime; the test
required them to live at different timestamps
though.

tschottdorf commented:
@spencerkimball your eyes on the `add PushNow to PushRequest` commit are appreciated. I think that allows us to completely get rid of explicit request timestamps (except when carried in a `Txn`).

tbg committed:
convert multiraft.ErrGroupDeleted early

and return *proto.Error from AddCmd

tschottdorf commented:
thanks @tamird, feedback is addressed (fixed most of it in the original commits, which unfortunately resulted in your comments being removed from the thread here).

tamird commented on client/batch.go:
```diff
@@ -194,20 +193,22 @@ func (b *Batch) fillResults(br *proto.BatchResponse, pErr *proto.Error) error {
-		*proto.IncrementRequest,
-		*proto.DeleteRequest:
-		numRows = 1
+// InternalAddRequest adds the specified request to the batch. It is intended
```
s/request/requests/

tschottdorf replied:
Done.

On Sun, Sep 27, 2015 at 2:39 PM, Tamir Duberstein <notifications@github.com>
wrote:

> In client/batch.go
> <https://github.com/cockroachdb/cockroach/pull/2678#discussion_r40507666>:
>
> > @@ -194,20 +193,22 @@ func (b *Batch) fillResults(br *proto.BatchResponse, pErr *proto.Error) error {
> >  	return nil
> >  }
> >
> > -// InternalAddRequest adds the specified call to the batch. It is intended for
> > -// internal use only.
> > -func (b *Batch) InternalAddRequest(args proto.Request) {
> > -	numRows := 0
> > -	switch args.(type) {
> > -	case *proto.GetRequest,
> > -		*proto.PutRequest,
> > -		*proto.ConditionalPutRequest,
> > -		*proto.IncrementRequest,
> > -		*proto.DeleteRequest:
> > -		numRows = 1
> > +// InternalAddRequest adds the specified request to the batch. It is intended
>
> s/request/requests/
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/2678/files#r40507666>.
>

tamird commented on kv/dist_sender.go:
```diff
@@ -444,7 +441,7 @@ func (ds *DistSender) sendAttempt(trace *tracer.Trace, ba proto.BatchRequest, de
 	}
 
 	// TODO(tschottdorf) &ba -> ba
```
this is still here

tschottdorf replied:
Whoops. Definitely got it now.

On Sun, Sep 27, 2015 at 2:40 PM, Tamir Duberstein <notifications@github.com>
wrote:

> In kv/dist_sender.go
> <https://github.com/cockroachdb/cockroach/pull/2678#discussion_r40507673>:
>
> > @@ -444,7 +441,7 @@ func (ds *DistSender) sendAttempt(trace *tracer.Trace, ba proto.BatchRequest, de
> >  	}
> >
> >  	// TODO(tschottdorf) &ba -> ba
>
> this is still here
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/2678/files#r40507673>.
>

tbg committed:
make PushTxnReq.PusherTxn non-nullable

with some minor fixes sprinkled in

tamird commented:
LGTM but wait for @spencerkimball

bdarnell commented on kv/txn_coord_sender.go:
```diff
@@ -322,8 +322,7 @@ func (tc *TxnCoordSender) Send(ctx context.Context, ba proto.BatchRequest) (*pro
 	defer trace.Finalize()
-	// TODO(tschottdorf): always "Batch"
-	defer trace.Epoch(fmt.Sprintf("sending %s", ba.Method()))()
+	defer trace.Epoch(fmt.Sprintf("sending %s", proto.Batch))()
```
I don't really like this change from `.Method()` to a hard-coded `proto.Batch` (here and elsewhere, especially where it's combined with a literal like in dist_sender_test.go). I'd prefer either keeping a `.Method()` method on `BatchRequest` (similar to the one we have on `sql/driver.Request`, even though the only method there is `Execute`), or dispense with the constant altogether and just make it a string literal: `trace.Epoch("sending Batch")`.

tschottdorf replied:
I've replaced uses of `proto.Batch` almost everywhere. There remain, I think, two places where a TODO is in place to get rid of them already.

bdarnell commented:
LGTM

tbg committed:
remove (almost all) uses of proto.Batch

tbg committed:
disable txn auto gc

this could lead to spurious test failures:

* txn writes some intents, but not in a single
  round-trip
* concurrent reader runs into open intent, tries to push
* before it gets there, txn finishes with only local intents;
  txn record gets deleted
* push begins, does not find txn entry and fails
* potentially repeat forever or (if a push succeeds) until
  recreated txn entry times out.

I ran into this since it created an endless loop in
conjunction with buggy test code, but it's better to
disable this for now until we actually have the GC
story in place. My intuition is that we'll want to
limit this to single-roundtrip txns, whose intents
are never visible to anyone except the requests in
the batch (and even then we must be careful that
they are not collected as "skipped intents" in
some cases).

tbg committed:
multiTestContext: fix potential infinite loop

I ran into this in conjunction with the issue described in the
last commit: a Push would fail every time, but rpcSend would
continue querying further ranges, so the error reported back
to DistSender was always NotLeader or RangeNotFound, causing
infinite retries.

tschottdorf commented:
I added two more commits to deal with flaky `TestStoreRangeRebalance`: one that turns of txn auto-gc and the other that fixes up some code in the "multiplexing" test `rpcSender` we use for `multiStoreSender` (so that it bails if it hits the right range and gets an error; previously it'd just try the next which could lead to infinite retries if your last store wasn't the right one and would give you key mismatch errors)

tamird commented:
still LGTM

bdarnell commented:
LGTM

mrtracy commented:
Also LGTM
